extends CharacterBody2D

@export var mSpeed: float = 200.0
@onready var mSprite = $AnimatedSprite2D

var mDirection: int = 0
var mIsMoving: bool = false

func _process(delta):
	if mIsMoving:
		var velocity: Vector2 = GetDirection(mDirection) * mSpeed
		position += velocity * delta

# ì„œë²„ê°€ ë³´ë‚´ì¤€ ìœ„ì¹˜ ë³´ì •
func SetPosition(newPos: Vector2) -> void:
	position = newPos

# ë¡œì»¬ ìž…ë ¥ ê¸°ë°˜ ì´ë™ ì‹œ í˜¸ì¶œ (í”Œë ˆì´ì–´ ì§ì ‘ ì¡°ìž‘ìš©)
func UpdateMovement(dt):
	var velocity: Vector2 = GetDirection(mDirection) * mSpeed
	position += velocity * dt
	
func SetWalk():
	if mSprite:
		mSprite.play("walk")
		
		# ðŸ‘‰ ë°©í–¥ì— ë”°ë¼ flip_h ì„¤ì •
		match mDirection:
			0, 1, 7:  # ì™¼ìª½ ê´€ë ¨ ë°©í–¥
				mSprite.flip_h = true
			3, 4, 5:  # ì˜¤ë¥¸ìª½ ê´€ë ¨ ë°©í–¥
				mSprite.flip_h = false
		
func SetIdle():
	if mSprite:
		mSprite.play("idle")

# 8ë°©í–¥ ì²˜ë¦¬
func GetDirection(dir: int) -> Vector2:
	match dir:
		0: return Vector2(-1, 0)            # LL
		1: return Vector2(-1, -1).normalized()
		2: return Vector2(0, -1)            # UU
		3: return Vector2(1, -1).normalized()
		4: return Vector2(1, 0)             # RR
		5: return Vector2(1, 1).normalized()
		6: return Vector2(0, 1)             # DD
		7: return Vector2(-1, 1).normalized()
		_: return Vector2.ZERO
